subsys/net/lib/fota_download/src/fota_download.c

```diff
@@ -7,22 +7,250 @@
+		return -1;
+	}
+
+	LOG_DBG("Flash memory available: %u", resource);
+
+	len = sizeof(off);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			    &off, &len);
+	if (rc < 0) {
+		LOG_ERR("Offset request failed, errno %d", errno);
+		return -1;
+	}
+
+	LOG_DBG("Offset retrieved: %u", off);
+
+	return rc;
+}
+
+static nrf_dfu_err_t modem_error_get(void)
+{
+	int rc;
+	nrf_dfu_err_t err;
+	nrf_socklen_t len;
+
+	len = sizeof(err);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_ERROR,
+			    &err, &len);
+	if (rc) {
+		LOG_ERR("Unable to fetch modem error, errno %d", errno);
+	}
+
+	return err;
+}
+
+static int modem_fw_write(u8_t *buf, size_t len)
+{
+	int sent = send(modem_dfu_socket_fd, buf, len, 0);
+
+	if (sent < 0) {
+		int modem_error = modem_error_get();
+
+		LOG_ERR("Modem refused fragment, errno %d, dfu err %d", errno,
+			modem_error);
+		if (modem_error == DFU_ERASE_PENDING) {
+			return -EAGAIN;
+		}
+		return -1;
+	}
+	return 0;
+}
+
+static void modem_fw_delete_task(struct k_work *w)
+{
+	int rc;
+	u32_t off;
+	nrf_socklen_t len;
+
+	len = sizeof(off);
+	rc = nrf_getsockopt(modem_dfu_socket_fd, NRF_SOL_DFU, NRF_SO_DFU_OFFSET,
+			    &off, &len);
+	if (rc < 0) {
+		k_delayed_work_submit_to_queue(&modem_fw_delete_offload_work_q,
+					       &modem_fw_delete_dwork,
+					       K_MSEC(500));
+	} else {
+		LOG_INF("Firmware image deleted");
+		k_sem_give(&modem_fw_delete_sem);
+	}
+}
+
+static int modem_dfu_socket(void)
+{
+	modem_dfu_socket_fd = nrf_socket(NRF_AF_LOCAL, NRF_SOCK_STREAM,
+				     NRF_PROTO_DFU);
+	if (modem_dfu_socket_fd < 0) {
+		LOG_ERR("Failed to open modem DFU socket.");
+		return -1;
+	}
+	return 0;
+}
+
+static int modem_dfu_socket_close(void)
+{
+	int rc = nrf_close(modem_dfu_socket_fd);
+
+	if (rc < 0) {
+		LOG_ERR("Failed to open modem DFU socket.");
+		return -1;
+	}
+
+	return 0;
+}
+#endif
+
+
+#define MCUBOOT  1
+#define MODEM_FW 2
+
+static inline void log_img_hdr(struct mcuboot_img_header_v2 *img_hdr)
+{
+	LOG_DBG("magic:\t\t0x%x\n", img_hdr->magic);
+	LOG_DBG("load address:\t0x%x\n", img_hdr->load_addr);
+	LOG_DBG("header size:\t%d\n", img_hdr->hdr_size);
+	LOG_DBG("pad1:\t\t%d\n", img_hdr->_pad1);
+	LOG_DBG("image_size:\t%d\n", img_hdr->img_size);
+	LOG_DBG("flags:\t0x%x\n", img_hdr->flags);
+	struct mcuboot_img_sem_ver ver = img_hdr->sem_ver;
+
+	LOG_DBG("image version:\t%d.%d.%d.%d\n", ver.major, ver.minor,
+	       ver.revision, ver.build_num);
+}
+
+
+
+static int mcuboot_write(struct flash_img_context *flash_img,
+			 u8_t *buf, size_t len)
+{
+	return flash_img_buffered_write(flash_img, buf, len, false);
+}
+
+static int check_img_type(u8_t *buf, size_t len)
+{
+	struct mcuboot_img_header_v2 *img_hdr = (struct mcuboot_img_header_v2 *)
+						buf;
+	if (len < BOOT_HEADER_SZ) {
+		return -EAGAIN;
+	}
+	if (img_hdr->magic == BOOT_MAGIC) {
+		log_img_hdr(img_hdr);
+		return MCUBOOT;
+	}
+#ifdef CONFIG_FOTA_DOWNLOAD_MODEM_FW_UPDATE_SUPPORT
+	else {
```

Comment:
I could not find any specification for the format of this and I'm unsure if we want to expose it. 

The clue here is that the modem would return an error with type `DFU_INVALID_UUID` if the magic + signature can't be verified by the modem. Which means we inherently trust the modem to do the right thing and the security of the modem is so that anything passed to it won't brick/compromise it.

Reply:
Would prefer to check for some magic if it exists @lemrey do you know if we want to expose this?

Reply:
Found the magic added.